src/Sets/Zonotope.jl

```diff
@@ -427,3 +428,65 @@ function linear_map!(Zout::Zonotope, M::AbstractMatrix, Z::Zonotope)
     mul!(Zout.generators, M, Z.generators)
     return Zout
 end
+
+"""
+    quadratic_map(Q::Vector{MT}, Z::Zonotope{N}) where {N, MT<:AbstractMatrix{N}}
+
+Return an overapproximation of the quadratic map of the given zonotope.
+
+### Input
+
+- `Z` -- zonotope
+- `Q` -- array of square matrices
+
+### Output
+
+An overapproximation of the quadratic map of the given zonotope.
+
+### Notes
+
+Mathematically, a quadratic map of a zonotope is defined as:
+
+```math
+Z_Q = \\right\\{ \\lambda | \\lambda_i = x^T Q\\^{(i)} x,~i = 1, \\ldots, n,~x \\in Z \\left\\}
+```
+such that each coordinate ``i`` of the resulting zonotope is influenced by ``Q\\^{(i)}``
+
+### Algorithm
+
+This function implements [Lemma 1, 1].
+
+[1] *Matthias Althoff and Bruce H. Krogh. 2012. Avoiding geometric intersection
+operations in reachability analysis of hybrid systems. In Proceedings of the
+15th ACM international conference on Hybrid Systems: Computation and Control
+(HSCC ’12). Association for Computing Machinery, New York, NY, USA, 45–54.*
+"""
+function quadratic_map(Q::Vector{MT}, Z::Zonotope{N}) where {N, MT<:AbstractMatrix{N}}
+    @assert length(Q) == dim(Z) "the number of matrices needs to match the dimension of the zonotope"
+    G = genmat(Z)
+    c = center(Z)
+    n, p = size(G)
+    h = Matrix{N}(undef, n, binomial(p+2, 2)-1)
+    d = Vector{N}(undef, n)
+    g(x) = view(G, :, x)
+    cᵀ = c'
+    for (i, Qᵢ) in enumerate(Q)
+        cᵀQᵢ = cᵀ * Qᵢ
+        Qᵢc = Qᵢ * c
+        aux = zero(N)
+        for j=1:p
+            aux += g(j)' * Qᵢ * g(j)
+            h[i, j] = cᵀQᵢ * g(j) + g(j)' * Qᵢc
+            h[i, p+j] = 0.5 * g(j)' * Qᵢ * g(j)
+        end
+        d[i] = cᵀQᵢ * c + 0.5 * aux
+        l = 0
+        for j=1:p-1
+            for k=j+1:p
+                l += 1
+                h[i, 2p+l] = g(j)' * Qᵢ * g(k) + g(k)' * Qᵢ * g(j)
+            end
+        end
```

Comment:
Here `g(j)' * Qᵢ` and `Qᵢ * g(j)` are constant in the inner loop, so you can pull them out.

Reply:
On second thought, is it actually correct to pull out the second expression? :sweat_smile:
Do you have to evaluate the product from left to right?

Reply:
i think it's the same; these products are associative